Introduce top-document and child-document content blocking resource types

elijahsawyers opened the pull request:
#### d5d74427397d8dec2e03456dae6ea4de50060d53
<pre>
Introduce top-document and child-document content blocking resource types
<a href="https://bugs.webkit.org/show_bug.cgi?id=290067">https://bugs.webkit.org/show_bug.cgi?id=290067</a>
<a href="https://rdar.apple.com/147442279">rdar://147442279</a>

Reviewed by Timothy Hatcher and Alex Christensen.

This patch breaks out the Document resource type to be TopDocument and
ChildDocument. The motive behind this change is that currently, a dNR rule that
doesn&apos;t specify resource types gets converted into two WebKit content blocking
rules.

E.g.

{
  ...

  trigger = {
    load-context = child-frame
    resource-type = document
  }

  ...
}

{
  ...

  trigger = {
    resource-type = script, ping, ...
  }

  ...
}

By introducing the child-document resource type, those two rules can be
collapsed into one.

* LayoutTests/http/tests/contentextensions/block-child-document-resource-type-expected.txt: Added.
* LayoutTests/http/tests/contentextensions/block-child-document-resource-type.html: Added.
* LayoutTests/http/tests/contentextensions/block-child-document-resource-type.html.json: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type-expected.txt: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type-with-child-document-expected.txt: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type-with-child-document.html: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type-with-child-document.html.json: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type.html: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type.html.json: Added.
Introduce new layout tests to verify that the top-document and child-document
resource types behave as expected.

* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
Update the resourceType check to look for TopDocument or ChildDocument, rather
than Document.

* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::subresourcesAllowReuse const):
Pass in whether or not the frame is the main frame.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::handleSubstituteDataLoadNow):
Set the ResourceType to either TopDocument or ChildDocument, rather than
Document.

* Source/WebCore/loader/ResourceLoadInfo.cpp:
(WebCore::ContentExtensions::toResourceType):
Return an option set with either TopDocument or ChildDocument ResourceType if
the cached resource type is a main resource. Also, update this method to take
in a parameter as to whether or not the resource is the main frame, which is
used to decide whether to return TopDocument or ChildDocument.

(WebCore::ContentExtensions::readResourceType):
Return an option set with both TopDocument and ChildDocument ResourceTypes if
the content blocking rule specifies &quot;document&quot; resource type.

Return TopDocument and ChildDocument for their resource type strings.

(WebCore::ContentExtensions::resourceTypeToString):
Return the resource type strings for the TopDocument and ChildDocument
ResourceTypes.

* Source/WebCore/loader/ResourceLoadInfo.h:
Replace the Document ResourceType with TopDocument and ChildDocument, and
update the byte code.

* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::setDocumentURL):
Set the ResourceType to either TopDocument or ChildDocument, rather than
Document.

* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::SubresourceLoader):
Pass in whether or not the frame is the main frame.

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
Pass in whether or not the frame is the main frame.

* Source/WebCore/page/UserContentProvider.cpp:
(WebCore::UserContentProvider::processContentRuleListsForLoad):
Update the resourceType check to look for TopDocument or ChildDocument, rather
than Document.

* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
Update the content rule list file version since we modified the byte code.

* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::mainDocumentRequest):
(TestWebKitAPI::subResourceRequest):
(TestWebKitAPI::requestInTopAndFrameURLs):
(TestWebKitAPI::TEST_F(ContentExtensionTest, TopURL)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, LoadType)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, ResourceType)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, ResourceOrLoadTypeMatchingEverything)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, MatchesEverything)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, IfFrameURL)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, UnlessFrameURL)):
Update the tests to use TopDocument and ChildDocument ResourceTypes, since Document
was removed.

Canonical link: <a href="https://commits.webkit.org/292482@main">https://commits.webkit.org/292482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08a644db08ff7c6ef7dba159c2093ef3cadfdcdf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96207 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15821 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5788 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/101272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/46726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98252 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16117 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24254 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/101272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/46726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99210 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/12090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/86912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/101272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/11841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/4673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/46052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/4770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/103300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23274 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/103300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23525 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/82931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/103300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/26377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/3815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15478 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23237 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22896 "Failed to checkout and rebase branch from PR 42722") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/26376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24637 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (approved):
LGTM, but I think it needs a test that verifies that if you use child-document the action will apply to an iframe load but not the main frame load and vice versa.

xeenon commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -242,7 +242,7 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
         if (initiatingDocumentLoader.isLoadingMainResource()
             && frame->isMainFrame()
-            && resourceType == ResourceType::Document)
+            && resourceType.containsAny({ ContentExtensions::ResourceType::TopDocument, ContentExtensions::ResourceType::ChildDocument }))
```
```suggestion
            && resourceType.containsAny({ ResourceType::TopDocument, ResourceType::ChildDocument }))
```

xeenon commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -270,7 +270,7 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
                 result.blockedCookies = true;
             }, [&](const CSSDisplayNoneSelectorAction& actionData) {
-                if (resourceType == ResourceType::Document)
+                if (resourceType.containsAny({ ContentExtensions::ResourceType::TopDocument, ContentExtensions::ResourceType::ChildDocument }))
```
```suggestion
                if (resourceType.containsAny({ ResourceType::TopDocument, ResourceType::ChildDocument }))
```

xeenon commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -300,7 +300,7 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
         if (!actionsFromContentRuleList.sawIgnorePreviousRules) {
             if (auto* styleSheetContents = globalDisplayNoneStyleSheet(contentRuleListIdentifier)) {
-                if (resourceType == ResourceType::Document)
+                if (resourceType.containsAny({ ContentExtensions::ResourceType::TopDocument, ContentExtensions::ResourceType::ChildDocument }))
```
```suggestion
                if (resourceType.containsAny({ ResourceType::TopDocument, ResourceType::ChildDocument }))
```

xeenon commented on Source/WebCore/loader/ResourceLoadInfo.cpp:
```diff
@@ -42,12 +42,15 @@ static_assert(!(LoadContextMask & ActionConditionMask), "LoadContextMask and Act
+        if (isMainFrame)
+            return { ResourceType::TopDocument };
+
+        return { ResourceType::ChildDocument };
```
```suggestion
            return { ResourceType::TopDocument };
        return { ResourceType::ChildDocument };
```

xeenon commented on Source/WebCore/loader/ResourceLoadInfo.h:
```diff
@@ -35,52 +35,53 @@ namespace WebCore::ContentExtensions {
+    Fetch = 0x0400,
+    WebSocket = 0x0800,
+    Other = 0x1000,
+    CSPReport = 0x2000,
```
Since we are changing values, Other makes most sense at the end.
```suggestion
    CSPReport = 0x1000,
    Other = 0x2000,
```

xeenon commented on Source/WebCore/loader/ResourceLoadInfo.h:
```diff
@@ -35,52 +35,53 @@ namespace WebCore::ContentExtensions {
+static constexpr uint64_t ActionFlagMask = 0x001FFFFF00000000;
 
-OptionSet<ResourceType> toResourceType(CachedResource::Type, ResourceRequestRequester);
+OptionSet<ResourceType> toResourceType(CachedResource::Type, ResourceRequestRequester, bool);
```
```suggestion
OptionSet<ResourceType> toResourceType(CachedResource::Type, ResourceRequestRequester, bool isMainFrame);
```

elijahsawyers committed:
Introduce top-document and child-document content blocking resource types
https://bugs.webkit.org/show_bug.cgi?id=290067
rdar://147442279

Reviewed by Timothy Hatcher and Alex Christensen.

This patch breaks out the Document resource type to be TopDocument and
ChildDocument. The motive behind this change is that currently, a dNR rule that
doesn't specify resource types gets converted into two WebKit content blocking
rules.

E.g.

{
  ...

  trigger = {
    load-context = child-frame
    resource-type = document
  }

  ...
}

{
  ...

  trigger = {
    resource-type = script, ping, ...
  }

  ...
}

By introducing the child-document resource type, those two rules can be
collapsed into one.

* LayoutTests/http/tests/contentextensions/block-child-document-resource-type-expected.txt: Added.
* LayoutTests/http/tests/contentextensions/block-child-document-resource-type.html: Added.
* LayoutTests/http/tests/contentextensions/block-child-document-resource-type.html.json: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type-expected.txt: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type-with-child-document-expected.txt: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type-with-child-document.html: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type-with-child-document.html.json: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type.html: Added.
* LayoutTests/http/tests/contentextensions/block-top-document-resource-type.html.json: Added.
Introduce new layout tests to verify that the top-document and child-document
resource types behave as expected.

* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
Update the resourceType check to look for TopDocument or ChildDocument, rather
than Document.

* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::subresourcesAllowReuse const):
Pass in whether or not the frame is the main frame.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::handleSubstituteDataLoadNow):
Set the ResourceType to either TopDocument or ChildDocument, rather than
Document.

* Source/WebCore/loader/ResourceLoadInfo.cpp:
(WebCore::ContentExtensions::toResourceType):
Return an option set with either TopDocument or ChildDocument ResourceType if
the cached resource type is a main resource. Also, update this method to take
in a parameter as to whether or not the resource is the main frame, which is
used to decide whether to return TopDocument or ChildDocument.

(WebCore::ContentExtensions::readResourceType):
Return an option set with both TopDocument and ChildDocument ResourceTypes if
the content blocking rule specifies "document" resource type.

Return TopDocument and ChildDocument for their resource type strings.

(WebCore::ContentExtensions::resourceTypeToString):
Return the resource type strings for the TopDocument and ChildDocument
ResourceTypes.

* Source/WebCore/loader/ResourceLoadInfo.h:
Replace the Document ResourceType with TopDocument and ChildDocument, and
update the byte code.

* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::setDocumentURL):
Set the ResourceType to either TopDocument or ChildDocument, rather than
Document.

* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::SubresourceLoader):
Pass in whether or not the frame is the main frame.

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
Pass in whether or not the frame is the main frame.

* Source/WebCore/page/UserContentProvider.cpp:
(WebCore::UserContentProvider::processContentRuleListsForLoad):
Update the resourceType check to look for TopDocument or ChildDocument, rather
than Document.

* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
Update the content rule list file version since we modified the byte code.

* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::mainDocumentRequest):
(TestWebKitAPI::subResourceRequest):
(TestWebKitAPI::requestInTopAndFrameURLs):
(TestWebKitAPI::TEST_F(ContentExtensionTest, TopURL)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, LoadType)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, ResourceType)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, ResourceOrLoadTypeMatchingEverything)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, MatchesEverything)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, IfFrameURL)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, UnlessFrameURL)):
Update the tests to use TopDocument and ChildDocument ResourceTypes, since Document
was removed.

Canonical link: https://commits.webkit.org/292482@main
